### DISCUSSION TITLE
Problem compiling building osgEarth on Red Hat Linux 7.9

### DISCUSSION THREAD

**Discussion by faculaganymede** [pelicanmapping/osgearth#2210] [Support Q&A]
osgEarth Gurus,

I need help with compiling and building osgEarth 3.3 on Red Hat Linux 7.9.

Following the instructions on the osgEarth website, I successfully installed the osgEarth SDK using vcpkg. Next, I downloaded the source code from GitHub and tried to compile and build it but ran into the following errors. I didn't modify any of the cmake files and assumed that they would work out of the box.

```
vcpkg install osgearth:x64-linux

git clone --recurse-submodules https://github.com/gwaldron/osgearth.git osgearth
cd osgearth
mkdir build
cd build

cmake -S .. -B . -DCMAKE_BUILD_TYPE=RelWithDebInfo -DWIN32_USE_MP=ON -DCMAKE_INSTALL_PREFIX=/home/facula/osgearth -DCMAKE_TOOLCHAIN_FILE=/home/facula/vcpkg/scripts/buildsystems/vcpkg.cmake

cmake --build . --config RelWithDebInfo

[ 53%] Built target osgEarth
[ 53%] Built target osgdb_basis
[ 53%] Built target osgdb_osgearth_bumpmap
[ 54%] Built target osgdb_osgearth_cache_filesystem
[ 54%] Built target osgdb_osgearth_colorramp
[ 55%] Built target osgdb_osgearth_detail
[ 55%] Built target osgdb_draco
[ 56%] Built target osgdb_earth
[ 60%] Built target osgdb_osgearth_engine_rex
[ 60%] Built target osgdb_osgearth_featurefilter_intersect
[ 60%] Built target osgdb_osgearth_featurefilter_join
[ 60%] Built target osgdb_gltf
[ 65%] Built target osgdb_kml
[ 66%] Built target osgdb_osgearth_mapinspector
[ 66%] Built target osgdb_osgearth_monitor
[ 67%] Built target osgdb_osgearth_scriptengine_javascript
[ 68%] Built target osgdb_osgearth_sky_gl
[ 69%] Built target osgdb_osgearth_sky_simple
[ 70%] Built target osgdb_template
[ 71%] Built target osgdb_osgearth_terrainshader
[ 71%] Built target osgdb_webp
[ 73%] Built target osgdb_lerc
[ 73%] Built target osgdb_osgearth_vdatum_egm2008
[ 73%] Built target osgdb_osgearth_vdatum_egm84
[ 74%] Built target osgdb_osgearth_vdatum_egm96
[ 74%] Built target osgdb_osgearth_viewpoints
[ 75%] Built target osgdb_fastdxt
[ 75%] Linking CXX executable ../../../bin/osgearth_viewer
../../../lib/libosgEarth.so.3.3.0: undefined reference to `FcDefaultSubstitute'
../../../lib/libosgEarth.so.3.3.0: undefined reference to `FcPatternDestroy'
../../../lib/libosgEarth.so.3.3.0: undefined reference to `LZ4_compress_HC'
../../../lib/libosgEarth.so.3.3.0: undefined reference to `shm_unlink'
../../../lib/libosgEarth.so.3.3.0: undefined reference to `LZ4_compress_fast'
../../../lib/libosgEarth.so.3.3.0: undefined reference to `snappy_compress'
../../../lib/libosgEarth.so.3.3.0: undefined reference to `FcNameParse'
../../../lib/libosgEarth.so.3.3.0: undefined reference to `LZ4_decompress_safe'
../../../lib/libosgEarth.so.3.3.0: undefined reference to `FcInitLoadConfigAndFonts'
../../../lib/libosgEarth.so.3.3.0: undefined reference to `FcFontMatch'
../../../lib/libosgEarth.so.3.3.0: undefined reference to `shm_open'
../../../lib/libosgEarth.so.3.3.0: undefined reference to `FcConfigSubstitute'
../../../lib/libosgEarth.so.3.3.0: undefined reference to `FcPatternGetString'
../../../lib/libosgEarth.so.3.3.0: undefined reference to `snappy_uncompress'
../../../lib/libosgEarth.so.3.3.0: undefined reference to `snappy_max_compressed_length'
collect2: error: ld returned 1 exit status
gmake[2]: *** [bin/osgearth_viewer] Error 1
gmake[1]: *** [src/applications/osgearth_viewer/CMakeFiles/application_osgearth_viewer.dir/all] Error 2
gmake: *** [all] Error 2
```

**Comment by jasonbeverage** [comment]
You shouldn't need to install osgearth from vcpkg and then try to build it from source, in theory it should build everything you need via vcpkg in manifest mode automatically.  

That being said I've never tried building with vcpkg on Linux, I usually prefer to use the system package manager to install deps on Linux.  Vcpkg is mostly useful for windows for me.  

It seems like maybe its having issues with either blocs or rocksbd on redhat.  Maybe try commenting those out in the vcpkg.json and see if you get further.

**Reply by faculaganymede** [reply]
Thanks very much for your reply, Jason.  Unfortunately, removing rocksbd from vcpkg.json didn't help, still get the same errors.  There's no blocs, but removing blosc (in case it's a typo) also didn't help.

I am able to install and run osgEarth successfully on Windows 10, but I've been having tremendous trouble installing and compiling osgEarth on Red Hat 7.9. 

After much effort on Red Hat 7.9, I finally got `vcpkg install osgearth:x64-linux` to install osgEarth successfully.  However, vcpkg didn't install the shared/dynamic libraries (no *.so files, only static libraries *.a files).  I tried to compile the osgEarth example applications, but nothing got created.  `make` in the following commands didn't return anything, and none of the example applications actually got compiled/built (no executable created).

```
cd /home/facula/vcpkg/buildtrees/osgearth/src/05a55a9c16-4650d25569.clean/src/applications
cmake .
make
```

For this reason, I downloaded the osgEarth source code via git, and then I came across the compile/build errors described in the above message.  (It maybe worth mentioning, I'm able to compile and run OpenSceneGraph code on this Red Hat computer.)

**Reply by faculaganymede** [reply]
Unlike in Windows, `vcpkg install osgearth:x64-linux` doesn't install any shared/dynamic libraries on Red Hat 7.9.  Is that how it's supposed to be?

My ultimate goal is to compile my own osgEarth program but I am having a lot of trouble, maybe not linking all the libraries correctly.  Thus, I'm trying to compile the example applications that came in the osgEarth installation, assuming those examples have the correct setups in the cmake files, but so far I can't even compile the examples.

**Reply by plevy** [reply]
I'd take a look at what Jason mentioned.   Either manifest mode with vcpkg but you may end up with the same lib issue. Otherwise use a package manager to install the dependencies and cmake should find the system installed dependencies.

**Reply by faculaganymede** [reply]
Thank you plevy!  I'm still trying...
